### PR TITLE
#251 Impersonation Performance Tune

### LINKED ISSUE #251: 💙 - UWP Impersonation Performance Tune
> 🟥🟩 **WINDOWS 10** - Version 1903+ is required for the UWP UI.
> 🟦🟨 [How to setup UWP](https://docs.microsoft.com/en-us/windows/uwp/get-started/get-set-up)

## Summary
Performance tune ImpersonationChatPageViewModel

## How to implement
Adjust the order of execution after Client_MessageReceivedHandler is triggered
Move the MessageText(TextBox) update to the SendMessageAsync event handler
Remove unused _channel update async from MessageReceived event handler.

## Interested? 😊 

If you'd like to help us by solving this issue, just say so in the comments. ❤️ 
And don't shy away from asking any kind of question you might have.

Should you have any questions at all, feel free to contact either me (@petrspelos) or @DraxCodes on our Discord server:

[Discord](https://discord.gg/cGhEZuk)

_And of course, thank you so much for contributing! 🎉 It's because of amazing people like yourself that Miunie exists._

### PR DESCRIPTION
**Pull request by 1n5an1ty**
## Related Issue

Fixes #251 

## Changes
Adjust the order of execution after Client_MessageReceivedHandler is triggered
Move the MessageText(TextBox) update to the SendMessageAsync event handler
Remove unused _channel update async from MessageReceived event handler.
Rename LoadMessages => LoadMessagesAsync

## Expected Feedback
Please test perf. on your local machine as it appears to act a little differently based on geographical location and hardware, and basic 5an1ty checks.

### CODE REVIEW

**Comment by 1n5an1ty** [comment]
It should be good to go now, but I'll wait to hear back from you.
I put the check in the .where clause so we are not doing any unnecessary casting. Let me know if you see anything wrong with that.

File: src/Miunie.Discord/Impersonation.cs [hunk 1]
```diff
@@ -31,41 +31,48 @@ public IEnumerable<GuildView> GetAvailableGuilds()
...
+                .Cast<SocketTextChannel>()
+                .Select(channel => new TextChannelView
                 {
-                    _logger.Log($"Miunie cannot read from the '{channel.Name}' channel.");
-                }
-            }
+                    Id = channel.Id,
+                    Name = $"# {channel.Name}",
+                    Messages = new MessageView[0]
+                });
```
Comment (petrspelos) [inline src/Miunie.Discord/Impersonation.cs]: So if we're doing this, I'd suggest extracting the content of the `Where` clause into a method as well as the contents of this `Select` clause.

Something like
`.Where(IsViewableTextChannel)`
and
`Select(ToTextChannelView)`

That would read really well! 😊

File: src/Miunie.Discord/Impersonation.cs [hunk 2]
```diff
@@ -31,41 +31,48 @@ public IEnumerable<GuildView> GetAvailableGuilds()
...
         }
 
         public async Task<IEnumerable<MessageView>> GetMessagesFromTextChannelAsync(ulong guildId, ulong channelId)
         {
+            if (guildId == 0) { return new MessageView[0]; }
+
             var guild = _discord.Client.GetGuild(guildId);
-            var textChannel = guild.Channels.Where(c => c is SocketTextChannel && c.Id == channelId).Cast<SocketTextChannel>().FirstOrDefault();
+            var textChannel = guild.Channels
+                .Where(c => c.Id == channelId && c is SocketTextChannel && c.GetUser(_discord.Client.CurrentUser.Id).GetPermissions(c).ViewChannel)
```
Comment (petrspelos) [inline src/Miunie.Discord/Impersonation.cs]: Same here, let's extract this expression into a well-names method. 😊

Comment (petrspelos) [reply src/Miunie.Discord/Impersonation.cs]: We might even reuse the previous method if added into this expression. Now that's DRY. 😅

**Pull request merged**